v2/event/event_unmarshal_test.go

```diff
@@ -0,0 +1,899 @@
+		body    []byte
+		want    *event.Event
+		wantErr error
+	}{
+		"struct data v0.3": {
+			body: mustJsonMarshal(t, map[string]interface{}{
+				"specversion":     "0.3",
+				"datacontenttype": "application/json",
+				"data": map[string]interface{}{
+					"a": 42,
+					"b": "testing",
+				},
+				"id":        "ABC-123",
+				"time":      now.Format(time.RFC3339Nano),
+				"type":      "com.example.test",
+				"exbool":    true,
+				"exint":     42,
+				"exstring":  "exstring",
+				"exbinary":  "AAECAw==",
+				"exurl":     "http://example.com/source",
+				"extime":    now.Format(time.RFC3339Nano),
+				"schemaurl": "http://example.com/schema",
+				"source":    "http://example.com/source",
+			}),
+			want: &event.Event{
+				Context: event.EventContextV03{
+					Type:            "com.example.test",
+					Source:          *source,
+					SchemaURL:       schema,
+					ID:              "ABC-123",
+					Time:            &now,
+					DataContentType: event.StringOfApplicationJSON(),
+					Extensions: map[string]interface{}{
+						"exbool":   true, // Boolean should be preserved
+						"exint":    int32(42),
+						"exstring": "exstring",
+						// Since byte, url and time are encoded as string, the unmarshal should just convert them to string
+						"exbinary": "AAECAw==",
+						"exurl":    "http://example.com/source",
+						"extime":   now.Format(time.RFC3339Nano),
+					},
+				}.AsV03(),
+				DataEncoded: mustJsonMarshal(t, DataExample{
+					AnInt:   42,
+					AString: "testing",
+				}),
+			},
+		},
+		"string data v0.3": {
+			body: mustJsonMarshal(t, map[string]interface{}{
+				"specversion":     "0.3",
+				"datacontenttype": "application/json",
+				"data":            "This is a string.",
+				"id":              "ABC-123",
+				"time":            now.Format(time.RFC3339Nano),
+				"type":            "com.example.test",
+				"exbool":          true,
+				"exint":           42,
+				"exstring":        "exstring",
+				"exbinary":        "AAECAw==",
+				"exurl":           "http://example.com/source",
+				"extime":          now.Format(time.RFC3339Nano),
+				"schemaurl":       "http://example.com/schema",
+				"source":          "http://example.com/source",
+			}),
+			want: &event.Event{
+				Context: event.EventContextV03{
+					Type:            "com.example.test",
+					Source:          *source,
+					SchemaURL:       schema,
+					ID:              "ABC-123",
+					Time:            &now,
+					DataContentType: event.StringOfApplicationJSON(),
+					Extensions: map[string]interface{}{
+						"exbool":   true, // Boolean should be preserved
+						"exint":    int32(42),
+						"exstring": "exstring",
+						// Since byte, url and time are encoded as string, the unmarshal should just convert them to string
+						"exbinary": "AAECAw==",
+						"exurl":    "http://example.com/source",
+						"extime":   now.Format(time.RFC3339Nano),
+					},
+				}.AsV03(),
+				DataEncoded: mustJsonMarshal(t, "This is a string."),
```

Comment:
Can you also add a test to validate we can deal with arrays in the data field? Arrays and strings were classic hard problems for the OG SDK

Reply:
A test for string data is already there, I'm adding one for array data